File: construction_v1/page/assistant_crm/assistant_crm.js
frappe.pages['assistant-crm'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Assistant CRM',
        single_column: true
    });

    // Initialize the page
    new AssistantCRMIntegration(page);
};

class AssistantCRMIntegration {
    constructor(page) {
        this.page = page;
        this.wrapper = page.main;
        this.init();
    }

    init() {
        // Assets are already loaded globally via hooks.py, no need for frappe.require()
        console.log("🔍 Assistant CRM page init - checking asset availability");
        console.log("ConstructionChatBubble class available:", typeof ConstructionChatBubble !== 'undefined');

        // Load the HTML content directly
        this.setupHTML();

        // Bind events
        this.bindEvents();

        // Load current settings
        this.loadSettings();

        // Initialize chat bubble if available
        setTimeout(() => {
            console.log("🔍 Assistant CRM page: Initializing chat bubble");
            console.log("ConstructionChatBubble class available:", typeof ConstructionChatBubble !== 'undefined');
            console.log("window.ConstructionChatBubble exists:", typeof window.ConstructionChatBubble !== 'undefined');
            console.log("Manual init function available:", typeof window.initializeConstructionChatBubble !== 'undefined');

            if (window.ConstructionChatBubble) {
                console.log("✅ Chat bubble exists, triggering auto-launch");
                // Chat bubble already exists, trigger auto-launch
                window.ConstructionChatBubble.autoLaunchCenterScreen();
            } else if (typeof ConstructionChatBubble !== 'undefined') {
                console.log("✅ Creating new chat bubble instance");
                // Create new chat bubble instance
                window.ConstructionChatBubble = new ConstructionChatBubble();
                // Auto-launch after creation
                setTimeout(() => {
                    if (window.ConstructionChatBubble && window.ConstructionChatBubble.autoLaunchCenterScreen) {
                        window.ConstructionChatBubble.autoLaunchCenterScreen();
                    }
                }, 500);
            } else if (typeof window.initializeConstructionChatBubble !== 'undefined') {
                console.log("✅ Using manual initialization function");
                window.initializeConstructionChatBubble();
                setTimeout(() => {
                    if (window.ConstructionChatBubble && window.ConstructionChatBubble.autoLaunchCenterScreen) {
                        window.ConstructionChatBubble.autoLaunchCenterScreen();
                    }
                }, 500);
            } else {
                console.log("❌ Cannot initialize chat bubble - class not available");
            }
        }, 1000);
    }

    setupHTML() {
        // Create HTML content without template issues
        const html = `
            <div class="assistant-crm-integration">
                <div class="page-header">
                    <div class="WorkCom-avatar">
                        <div class="WorkCom-image">👩‍💼</div>
                    </div>
                    <h1>Hi this is WorkCom!</h1>
                    <p class="WorkCom-greeting">How can I help?</p>
                    <p class="text-muted">Your Construction AI assistant for customer service and CRM support</p>
                </div>

                <!-- WorkCom's Services Section -->
                <div class="WorkCom-services-section">
                    <h3>I specialize in helping you with:</h3>
                    <div class="services-grid">
                        <div class="service-card">
                            <div class="service-icon">📋</div>
                            <h4>Claims Processing</h4>
                            <p>Assistance with workers' compensation claims and documentation</p>
                        </div>
                        <div class="service-card">
                            <div class="service-icon">🏢</div>
                            <h4>Employer Registration</h4>
                            <p>Help with company registration and compliance requirements</p>
                        </div>
                        <div class="service-card">
                            <div class="service-icon">💰</div>
                            <h4>Payment Services</h4>
                            <p>Support with online payments and declaration submissions</p>
                        </div>
                        <div class="service-card">
                            <div class="service-icon">📊</div>
                            <h4>Reports & Analytics</h4>
                            <p>Access to reports, statistics, and business insights</p>
                        </div>
                        <div class="service-card">
                            <div class="service-icon">🛡️</div>
                            <h4>Safety & Health</h4>
                            <p>Workplace safety guidelines and health compliance</p>
                        </div>
                        <div class="service-card">
                            <div class="service-icon">📞</div>
                            <h4>General Support</h4>
                            <p>Any questions about Construction services and procedures</p>
                        </div>
                    </div>
                </div>

                <div class="integration-content">
                    <!-- Enable/Disable Section -->
                    <div class="form-group">
                        <div class="checkbox">
                            <label>
                                <input type="checkbox" id="assistant-enabled" />
                                <span class="label-area">Enable WorkCom Assistant</span>
                            </label>
                        </div>
                        <p class="help-text text-muted">
                            Enable WorkCom to provide AI-powered assistance across your Construction system
                        </p>
                    </div>

                    <!-- Settings Form (hidden by default) -->
                    <div id="assistant-settings-form" style="display: none;">
                        <hr>
                        
                        <!-- AI Provider Section -->
                        <div class="form-group">
                            <label class="control-label">AI Provider</label>
                            <select class="form-control" id="ai-provider">
                                <option value="Google Gemini">Google Gemini</option>
                                <option value="OpenAI GPT">OpenAI GPT</option>
                                <option value="Anthropic Claude">Anthropic Claude</option>
                                <option value="Custom API">Custom API</option>
                            </select>
                        </div>

                        <!-- API Key Section -->
                        <div class="form-group">
                            <label class="control-label">API Key</label>
                            <input type="password" class="form-control" id="api-key" placeholder="Enter your AI provider API key">
                            <p class="help-text text-muted">
                                Your API key will be encrypted and stored securely
                            </p>
                        </div>

                        <!-- Model Name Section -->
                        <div class="form-group">
                            <label class="control-label">Model Name</label>
                            <input type="text" class="form-control" id="model-name" placeholder="e.g., gemini-1.5-pro">
                            <p class="help-text text-muted">
                                Specify the AI model to use (e.g., gemini-1.5-pro, gpt-4, claude-3-sonnet-20240229)
                            </p>
                        </div>

                        <!-- Advanced Settings -->
                        <div class="form-group">
                            <label class="control-label">Response Timeout (seconds)</label>
                            <input type="number" class="form-control" id="response-timeout" value="30" min="5" max="120">
                        </div>

                        <div class="form-group">
                            <label class="control-label">Welcome Message</label>
                            <textarea class="form-control" id="welcome-message" rows="2" placeholder="Hi! I'm WorkCom, your Construction assistant. How can I help you today? 😊"></textarea>
                        </div>

                        <div class="form-group">
                            <label class="control-label">Chat Bubble Position</label>
                            <select class="form-control" id="chat-position">
                                <option value="Bottom Right">Bottom Right</option>
                                <option value="Bottom Left">Bottom Left</option>
                                <option value="Top Right">Top Right</option>
                                <option value="Top Left">Top Left</option>
                            </select>
                        </div>

                        <!-- Action Buttons -->
                        <div class="form-group">
                            <button class="btn btn-primary" id="save-settings">Save Settings</button>
                            <button class="btn btn-default" id="test-connection" style="margin-left: 10px;">Test Connection</button>
                        </div>

                        <!-- Status Messages -->
                        <div id="status-message" style="margin-top: 15px;"></div>
                    </div>

                    <!-- Help Section -->
                    <div class="help-section" style="margin-top: 30px;">
                        <h4>Getting Started</h4>
                        <ol>
                            <li><strong>Get an API Key:</strong> Sign up for your preferred AI provider (Google AI Studio, OpenAI, Anthropic)</li>
                            <li><strong>Enable Integration:</strong> Check the "Enable Assistant CRM" checkbox above</li>
                            <li><strong>Configure Settings:</strong> Enter your API key and configure preferences</li>
                            <li><strong>Test Connection:</strong> Use the "Test Connection" button to verify your setup</li>
                            <li><strong>Start Using:</strong> The chat bubble will appear on all desk pages when properly configured</li>
                        </ol>

                        <h4>Features</h4>
                        <ul>
                            <li>🌐 <strong>Omnichannel Support:</strong> WhatsApp, Telegram, Facebook Messenger integration</li>
                            <li>🎯 <strong>CRM Integration:</strong> Direct access to customer data, leads, and opportunities</li>
                            <li>🤖 <strong>WorkCom Persona:</strong> Professional, patient, and solution-focused assistant</li>
                            <li>🔍 <strong>Context-Aware:</strong> Understands customer history and preferences</li>
                            <li>🛡️ <strong>Secure:</strong> Role-based access with encrypted API key storage</li>
                            <li>📊 <strong>Analytics:</strong> Sentiment analysis and conversation monitoring</li>
                        </ul>

                        <h4>Support</h4>
                        <p>For help and documentation, visit: <a href="https://github.com/QuantumSolver/construction_v1" target="_blank">Assistant CRM GitHub</a></p>
                    </div>
                </div>
            </div>
        `;

        $(this.wrapper).html(html);

        // Add debug controls
        this.addDebugControls();
    }

    addDebugControls() {
        // Add debug buttons to the page
        const debugHTML = `
            <div class="debug-controls" style="margin: 20px 0; padding: 15px; background: #f8f9fa; border-radius: 5px; border: 1px solid #dee2e6;">
                <h5 style="margin-bottom: 10px;">🔧 Debug Controls</h5>
                <button class="btn btn-primary btn-sm" onclick="window.testChatBubble()">Test Chat Bubble</button>
                <button class="btn btn-secondary btn-sm" onclick="window.checkAssets()">Check Assets</button>
                <button class="btn btn-info btn-sm" onclick="window.manualInit()">Manual Init</button>
                <button class="btn btn-warning btn-sm" onclick="window.forceAutoLaunch()">Force Auto Launch</button>
            </div>
        `;

        $(this.wrapper).prepend(debugHTML);

        // Add debug functions to window
        window.testChatBubble = () => {
            console.log("🧪 Testing chat bubble...");
            console.log("ConstructionChatBubble available:", typeof ConstructionChatBubble !== 'undefined');
            console.log("window.ConstructionChatBubble exists:", typeof window.ConstructionChatBubble !== 'undefined');
            if (window.ConstructionChatBubble) {
                console.log("Chat bubble methods:", Object.getOwnPropertyNames(Object.getPrototypeOf(window.ConstructionChatBubble)));
            }
        };

        window.checkAssets = () => {
            console.log("🔍 Checking assets...");
            console.log("jQuery available:", typeof $ !== 'undefined');
            console.log("Frappe available:", typeof frappe !== 'undefined');
            console.log("Frappe boot available:", frappe && typeof frappe.boot !== 'undefined');
            console.log("ConstructionChatBubble class:", typeof ConstructionChatBubble !== 'undefined');
        };

        window.manualInit = () => {
            console.log("🔧 Manual initialization...");
            if (typeof ConstructionChatBubble !== 'undefined') {
                window.ConstructionChatBubble = new ConstructionChatBubble();
                console.log("✅ Chat bubble created manually");
            } else {
                console.log("❌ ConstructionChatBubble class not available");
            }
        };

        window.forceAutoLaunch = () => {
            console.log("🚀 Force auto launch...");
            if (window.ConstructionChatBubble && window.ConstructionChatBubble.autoLaunchCenterScreen) {
                window.ConstructionChatBubble.autoLaunchCenterScreen();
                console.log("✅ Auto launch triggered");
            } else {
                console.log("❌ Chat bubble or auto launch method not available");
            }
        };
    }

    bindEvents() {
        const self = this;

        // Toggle settings form when enabled checkbox changes
        $(this.wrapper).on('change', '#assistant-enabled', function() {
            if ($(this).is(':checked')) {
                $('#assistant-settings-form').slideDown();
            } else {
                $('#assistant-settings-form').slideUp();
            }
        });

        // Save settings button
        $(this.wrapper).on('click', '#save-settings', function() {
            self.saveSettings();
        });

        // Test connection button
        $(this.wrapper).on('click', '#test-connection', function() {
            self.testConnection();
        });
    }

    loadSettings() {
        frappe.call({
            method: 'construction_v1.page.construction_v1.construction_v1.get_settings',
            callback: (r) => {
                if (r.message) {
                    this.populateForm(r.message);
                }
            }
        });
    }

    populateForm(settings) {
        $('#assistant-enabled').prop('checked', settings.enabled);
        $('#ai-provider').val(settings.ai_provider);
        $('#model-name').val(settings.model_name);
        $('#api-key').val(settings.has_api_key ? '***' : '');
        $('#response-timeout').val(settings.response_timeout);
        $('#welcome-message').val(settings.welcome_message);
        $('#chat-position').val(settings.chat_bubble_position);

        // Show/hide settings form based on enabled state
        if (settings.enabled) {
            $('#assistant-settings-form').show();
        }

        // Show status if there's an error
        if (settings.error) {
            this.showMessage('Error loading settings: ' + settings.error, 'red');
        }
    }

    saveSettings() {
        const settings = {
            enabled: $('#assistant-enabled').is(':checked'),
            ai_provider: $('#ai-provider').val(),
            model_name: $('#model-name').val(),
            api_key: $('#api-key').val(),
            response_timeout: parseInt($('#response-timeout').val()),
            welcome_message: $('#welcome-message').val(),
            chat_bubble_position: $('#chat-position').val()
        };

        // Validate required fields
        if (settings.enabled) {
            if (!settings.api_key || settings.api_key === '***') {
                this.showMessage('API Key is required when Assistant CRM is enabled', 'red');
                return;
            }
            if (!settings.model_name) {
                this.showMessage('Model Name is required', 'red');
                return;
            }
        }

        // Show loading
        this.showMessage('Saving settings...', 'blue');

        frappe.call({
            method: 'construction_v1.page.construction_v1.construction_v1.save_settings',
            args: {
                settings: settings
            },
            callback: (r) => {
                if (r.message && r.message.success) {
                    this.showMessage('Settings saved successfully! Chat bubble will appear when enabled.', 'green');

                    // Refresh the page to update chat bubble
                    setTimeout(() => {
                        window.location.reload();
                    }, 2000);
                } else {
                    this.showMessage('Error saving settings: ' + (r.message.error || 'Unknown error'), 'red');
                }
            }
        });
    }

    testConnection() {
        // Check if API key is provided
        const apiKey = $('#api-key').val();
        if (!apiKey || apiKey === '***') {
            this.showMessage('Please enter an API key before testing connection', 'red');
            return;
        }

        // Save settings first, then test
        this.showMessage('Testing connection...', 'blue');

        frappe.call({
            method: 'construction_v1.page.construction_v1.construction_v1.test_connection',
            callback: (r) => {
                if (r.message && r.message.success) {
                    this.showMessage(
                        `✅ Connection successful!<br>
                        <strong>Model:</strong> ${r.message.model}<br>
                        <strong>Response Preview:</strong> ${r.message.response_preview}`,
                        'green'
                    );
                } else {
                    this.showMessage(
                        `❌ Connection failed: ${r.message.message}<br>
                        <small>Please check your API key and model settings.</small>`,
                        'red'
                    );
                }
            }
        });
    }

    showMessage(message, color) {
        const alertClass = {
            'green': 'alert-success',
            'red': 'alert-danger',
            'blue': 'alert-info',
            'orange': 'alert-warning'
        }[color] || 'alert-info';

        $('#status-message').html(`
            <div class="alert ${alertClass}" role="alert">
                ${message}
            </div>
        `);

        // Auto-hide success messages after 5 seconds
        if (color === 'green') {
            setTimeout(() => {
                $('#status-message').fadeOut();
            }, 5000);
        }
    }
}

File: construction_v1/page/assistant_crm/assistant_crm.py
# Copyright (c) 2025, Assistant CRM and contributors
# For license information, please see license.txt

import frappe
from frappe import _


@frappe.whitelist()
def get_settings():
    """Get Assistant CRM settings for the integration page"""
    try:
        # Try to get existing settings - handle both Single and regular DocType
        try:
            settings = frappe.get_single("Assistant CRM Settings")
            return {
                "enabled": getattr(settings, 'enabled', False),
                "ai_provider": getattr(settings, 'ai_provider', 'Google Gemini'),
                "model_name": getattr(settings, 'model_name', 'gemini-1.5-pro'),
                "api_key": "***" if settings.get_password("api_key") else "",
                "has_api_key": bool(settings.get_password("api_key")),
                "response_timeout": getattr(settings, 'response_timeout', 30),
                "welcome_message": getattr(settings, 'welcome_message', 'Hello! I\'m WorkCom, your Construction assistant. How can I help you today?'),
                "chat_bubble_position": getattr(settings, 'chat_bubble_position', 'Bottom Right')
            }
        except Exception:
            # If Single DocType fails, try regular DocType approach
            if frappe.db.exists("Assistant CRM Settings", "Assistant CRM Configuration"):
                settings = frappe.get_doc("Assistant CRM Settings", "Assistant CRM Configuration")
                return {
                    "enabled": getattr(settings, 'enabled', False),
                    "ai_provider": getattr(settings, 'ai_provider', 'Google Gemini'),
                    "model_name": getattr(settings, 'model_name', 'gemini-1.5-pro'),
                    "api_key": "***" if getattr(settings, 'api_key', '') else "",
                    "has_api_key": bool(getattr(settings, 'api_key', '')),
                    "response_timeout": getattr(settings, 'response_timeout', 30),
                    "welcome_message": getattr(settings, 'welcome_message', 'Hello! I\'m WorkCom, your Construction assistant. How can I help you today?'),
                    "chat_bubble_position": getattr(settings, 'chat_bubble_position', 'Bottom Right')
                }
            else:
                # Return default settings if none exist
                return {
                    "enabled": False,
                    "ai_provider": "Google Gemini",
                    "model_name": "gemini-1.5-pro",
                    "api_key": "",
                    "has_api_key": False,
                    "response_timeout": 30,
                    "welcome_message": "Hello! I'm WorkCom, your Construction assistant. How can I help you today?",
                    "chat_bubble_position": "Bottom Right"
                }
    except Exception as e:
        frappe.log_error(f"Error getting Assistant CRM settings: {str(e)}", "Assistant CRM Page")
        # Return default settings on error
        return {
            "enabled": False,
            "ai_provider": "Google Gemini",
            "model_name": "gemini-1.5-pro",
            "api_key": "",
            "has_api_key": False,
            "response_timeout": 30,
            "welcome_message": "Hello! I'm WorkCom, your Construction assistant. How can I help you today?",
            "chat_bubble_position": "Bottom Right"
        }


@frappe.whitelist()
def save_settings(**kwargs):
    """Save Assistant CRM settings from the integration page"""
    try:
        import json

        # Handle different parameter formats for HTTP requests
        settings = kwargs.get('settings')

        # Try multiple sources for settings data
        if not settings:
            settings = frappe.form_dict.get('settings')

        if not settings:
            # Try to get individual parameters from form data
            form_data = frappe.form_dict
            if any(key in form_data for key in ['enabled', 'ai_provider', 'model_name', 'api_key']):
                settings = {
                    'enabled': form_data.get('enabled', False),
                    'ai_provider': form_data.get('ai_provider', 'Google Gemini'),
                    'model_name': form_data.get('model_name', 'gemini-1.5-pro'),
                    'api_key': form_data.get('api_key', ''),
                    'response_timeout': form_data.get('response_timeout', 30),
                    'welcome_message': form_data.get('welcome_message', 'Hello! How can I help you today?'),
                    'chat_bubble_position': form_data.get('chat_bubble_position', 'Bottom Right')
                }

        if isinstance(settings, str):
            settings = json.loads(settings)
        elif not settings:
            # If still no settings, return error with more details
            return {
                "success": False,
                "message": "No settings data provided",
                "debug_info": {
                    "kwargs": list(kwargs.keys()),
                    "form_dict": list(frappe.form_dict.keys())
                }
            }
        
        # Get or create settings document (Single DocType)
        try:
            doc = frappe.get_single("Assistant CRM Settings")
        except Exception:
            # If Single DocType fails, create new document
            doc = frappe.new_doc("Assistant CRM Settings")
            doc.name = "Assistant CRM Settings"  # Single DocType name
        
        # Update settings
        doc.enabled = settings.get("enabled", False)
        doc.ai_provider = settings.get("ai_provider", "Google Gemini")
        doc.model_name = settings.get("model_name", "gemini-1.5-pro")
        doc.response_timeout = settings.get("response_timeout", 30)
        doc.welcome_message = settings.get("welcome_message", "Hello! I'm WorkCom, your Construction assistant. How can I help you today?")
        doc.chat_bubble_position = settings.get("chat_bubble_position", "Bottom Right")
        
        # Handle API key
        if settings.get("api_key") and settings.get("api_key") != "***":
            doc.api_key = settings.get("api_key")
        
        # Save the document (Single DocType)
        try:
            doc.save()
        except Exception as e:
            # If save fails, try insert for new Single DocType
            if "does not exist" in str(e).lower():
                doc.insert()
            else:
                raise e
        
        return {"success": True, "message": "Settings saved successfully"}
        
    except Exception as e:
        frappe.log_error(f"Error saving Assistant CRM settings: {str(e)}", "Assistant CRM Page")
        return {"success": False, "error": str(e)}


@frappe.whitelist()
def test_connection():
    """Test the AI API connection"""
    try:
        from construction_v1.services.settings_service import get_settings_service
        
        settings_service = get_settings_service()
        ai_config = settings_service.get_ai_config()
        
        if not ai_config.get("api_key"):
            return {"success": False, "message": "API key is required"}
        
        # Test based on provider
        if ai_config.get("provider") == "Google Gemini":
            return test_gemini_connection(ai_config)
        else:
            return {"success": False, "message": "Provider not supported for testing yet"}
            
    except Exception as e:
        return {"success": False, "message": str(e)}


def test_gemini_connection(ai_config):
    """Test Google Gemini connection"""
    try:
        import google.generativeai as genai

        genai.configure(api_key=ai_config.get("api_key"))
        model = genai.GenerativeModel(ai_config.get("model_name", "gemini-1.5-pro"))
        response = model.generate_content("Test connection")

        return {
            "success": True,
            "message": "Connection successful",
            "model": ai_config.get("model_name"),
            "response_preview": response.text[:100] + "..." if len(response.text) > 100 else response.text
        }

    except Exception as e:
        return {"success": False, "message": f"Connection failed: {str(e)}"}


def get_context(context):
    """Get context for page template rendering"""
    try:
        # Add any context variables needed for the page template
        context.update({
            "app_name": "Assistant CRM",
            "app_version": "0.0.1",
            "page_title": "Assistant CRM Dashboard"
        })
        return context
    except Exception as e:
        frappe.log_error(f"Error getting page context: {str(e)}", "Assistant CRM Page")
        return context


